sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/AvroHDFSFileSource.java

```diff
@@ -0,0 +1,146 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.hdfs;
+
+import org.apache.beam.sdk.coders.AvroCoder;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.KvCoder;
+import org.apache.beam.sdk.io.BoundedSource;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.util.CoderUtils;
+import org.apache.beam.sdk.values.KV;
+
+import com.google.common.base.Function;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+
+import org.apache.avro.Schema;
+import org.apache.avro.mapred.AvroKey;
+import org.apache.avro.mapreduce.AvroJob;
+import org.apache.avro.mapreduce.AvroKeyInputFormat;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.mapreduce.InputSplit;
+import org.apache.hadoop.mapreduce.lib.input.FileInputFormat;
+
+import java.io.IOException;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * A {@code BoundedSource} for reading Avro files resident in a Hadoop filesystem.
+ *
+ * @param <T> The type of the Avro records to be read from the source.
+ */
+public class AvroHDFSFileSource<T> extends HDFSFileSource<AvroKey<T>, NullWritable> {
+  private static final long serialVersionUID = 0L;
+
+  protected final AvroCoder<T> avroCoder;
+  private final String schemaStr;
+
+  public AvroHDFSFileSource(String filepattern, AvroCoder<T> avroCoder) {
+    this(filepattern, avroCoder, null);
+  }
+
+  public AvroHDFSFileSource(String filepattern,
+                            AvroCoder<T> avroCoder,
+                            HDFSFileSource.SerializableSplit serializableSplit) {
+    super(filepattern,
+        ClassUtil.<AvroKeyInputFormat<T>>castClass(AvroKeyInputFormat.class),
+        ClassUtil.<AvroKey<T>>castClass(AvroKey.class),
+        NullWritable.class, serializableSplit);
+    this.avroCoder = avroCoder;
+    this.schemaStr = avroCoder.getSchema().toString();
+  }
+
+  @Override
+  public Coder<KV<AvroKey<T>, NullWritable>> getDefaultOutputCoder() {
+    AvroWrapperCoder<AvroKey<T>, T> keyCoder = AvroWrapperCoder.of(this.getKeyClass(), avroCoder);
+    return KvCoder.of(keyCoder, WritableCoder.of(NullWritable.class));
+  }
+
+  @Override
+  public List<? extends AvroHDFSFileSource<T>> splitIntoBundles(
+      long desiredBundleSizeBytes, PipelineOptions options) throws Exception {
+    if (serializableSplit == null) {
+      return Lists.transform(computeSplits(desiredBundleSizeBytes),
+          new Function<InputSplit, AvroHDFSFileSource<T>>() {
+            @Nullable
+            @Override
+            public AvroHDFSFileSource<T> apply(@Nullable InputSplit inputSplit) {
```

Comment:
Right it shouldn't be null. It's inherited from `Function<F, T>` and I kept it as is. Do you normally remove/change them to reflect the implementation?